File: python/pig-latin/pig_latin.py
def translate(text):
    text = text.split(' ')
    
    for i, word in enumerate(text):
        if word[0] not in ['a', 'e', 'i', 'o', 'u'] and \
           word[0:2] not in ['xr', 'yt']:
            word = shift(word)
    
        text[i] = f"{word}ay"

    return " ".join(text)

def shift(word):
    word = list(word)
    word.append(word.pop(0))

    if word[0] == 'u' and word[-1] == 'q':
        word.append(word.pop(0))
        return "".join(word)

    if word[0] in ['a', 'e', 'i', 'o', 'u', 'y']:
        return "".join(word)

    return shift("".join(word))
    

File: javascript/pig-latin/pig-latin.js
//
// This is only a SKELETON file for the 'Pig Latin' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const translate = (text) => {
  const words = text.split(' ')
  
  for (let [index, word] of words.entries()) {
    if (!['a', 'e', 'i', 'o', 'u'].includes(word[0]) && !['xr', 'yt'].includes(word.substring(0, 2))) {
        word = shift(word)
    }
    
    words[index] = `${word}ay`
    console.log(word, index, text)
  }

  return words.join(' ')
}

const shift = (word) => {
  word = word.split('')
  word.push(word.shift())

  if (word[0] == 'u' && word[word.length-1] == 'q') {
    word.push(word.shift())
    return word.join('')
  }

  if (['a', 'e', 'i', 'o', 'u', 'y'].includes(word[0])) {
    return word.join('')
  }

  return shift(word.join(''))
}